### PR TITLE
Typescript grammer

### PR DESCRIPTION
**Pull request by loonydev**
Adding grammar for Typescript.
I used JS grammer, combine it with [TypeScript spec](https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#A.1) and with some part from https://github.com/antlr/grammars-v4/issues/858

Yes, it's probably not coverage all.
Yes, need clear grammer, simplify and etc.
But it's work very well for my production code (400-500 lines of code) and it's good start.

### CODE REVIEW

File: typescript/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# TypeScript grammar
+
+This JTypeScript grammar does not exactly corresponds to TypeScript standard.
```
Comment (studentmain) [inline typescript/README.md]: JTypeScript?

**Comment by KvanTTT** [comment]
Thank you very much! You are the great men 👍

File: typescript/README.md (hunk 1 above)
Comment (loonydev) [reply typescript/README.md]: Yep, sorry, as I wrote, based on JS)

**Comment by loonydev** [comment]
> Nice work, suggest add some pure JS test cases.

About JS test cases, you can check with `javascript/examples/`, I'm not sure that it's good idea to copy inside this folder. Also I add targets from JS grammer later (Go, CShapr, Cpp, JS).

File: typescript/TypeScriptParser.g4 [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+enumDeclaration
+    : Const? Enum Identifier '{' enumBody? '}'
+    ;
+
+enumBody
+    :enumMemberList ','?
+    ;
+
+enumMemberList
+    :enumMember (',' enumMember)*
```
Comment (KvanTTT) [inline typescript/TypeScriptParser.g4]: Space after the comma.

File: typescript/TypeScriptParser.g4 [hunk 3]
```diff
@@ -0,0 +1,834 @@
...
+typeAliasDeclaration
+    : 'type' Identifier typeParameters? '=' type_ SemiColon
+    ;
+
+constructorDeclaration
+    : accessibilityModifier? Constructor '(' formalParameterList? ')' ( ('{' functionBody '}') | SemiColon)?
+    ;
+
+// A.5 Interface
+//
```
Comment (KvanTTT) [inline typescript/TypeScriptParser.g4]: Why do you use such an empty comment line?

File: typescript/TypeScriptParser.g4 [hunk 4]
```diff
@@ -0,0 +1,834 @@
...
+    : '{' statementList? '}'
+    ;
+
+statementList
+    : statement+
+    ;
+
+abstractDeclaration
+    : Abstract Identifier callSignature eos
+    | Abstract variableStatement eos
```
Comment (KvanTTT) [inline typescript/TypeScriptParser.g4]:
```antlr
: Abstract (Identifier callSignature | variableStatement) eos
```

File: typescript/TypeScriptParser.g4 [hunk 5]
```diff
@@ -0,0 +1,834 @@
...
+    | indexMemberDeclaration
+    | statement
+    ;
+
+propertyMemberDeclaration
+    : propertyMemberBase propertyName typeAnnotation? initializer? SemiColon
+    | propertyMemberBase propertyName callSignature ( ('{' functionBody '}') | SemiColon)
+    | propertyMemberBase (getAccessor | setAccessor)
+    | abstractDeclaration
+    ;
```
Comment (KvanTTT) [inline typescript/TypeScriptParser.g4]:
```antlr
propertyMemberDeclaration
    : propertyMemberBase (propertyName ( typeAnnotation? initializer? SemiColon
                                       | callSignature ( ('{' functionBody '}') | SemiColon))
                         | getAccessor
                         | setAccessor)
    | abstractDeclaration
    ;
```

File: typescript/TypeScriptParser.g4 [hunk 6]
```diff
@@ -0,0 +1,834 @@
...
+iteratorDefinition
+    : '[' singleExpression ']' '(' formalParameterList? ')' '{' functionBody '}'
+    ;
+
+formalParameterList
+    : formalParameterArg (',' formalParameterArg)* (',' lastFormalParameterArg)?
+    | lastFormalParameterArg
+    | arrayLiteral                            // ECMAScript 6: Parameter Context Matching
+    | objectLiteral                           // ECMAScript 6: Parameter Context Matching
+    | objectLiteral ':' formalParameterList   // ECMAScript 6: Parameter Context Matching
```
Comment (KvanTTT) [inline typescript/TypeScriptParser.g4]:
```antlr
| objectLiteral (':' formalParameterList)?   // ECMAScript 6: Parameter Context Matching
```

File: typescript/TypeScriptParser.g4 [hunk 7]
```diff
@@ -0,0 +1,834 @@
...
+// TSPart
+//
+// A.1 Types
+//
+
+typeParameters
+    : '<' typeParameterList? '>'
+    ;
+
+
```
Comment (KvanTTT) [inline typescript/TypeScriptParser.g4]: Try to always remove excess line breaks and use only one line break as a separator.

File: typescript/TypeScriptParser.g4 [hunk 8]
```diff
@@ -0,0 +1,834 @@
...
+    | predefinedType (',' predefinedType)*
+    ;
+
+requiredParameterList
+    : requiredParameter (',' requiredParameter)*
+    ;
+
+requiredParameter
+    : decoratorList? accessibilityModifier? identifierOrPattern typeAnnotation?
+    | decoratorList? identifierName ':' StringLiteral
```
[the remark this answers is not in the document]
Comment (loonydev) [reply typescript/TypeScriptParser.g4]: Hmmm `identifierOrPattern` include `identifierName`, `':' StringLiteral` it's some kind of type. So, I will include `StringLiteral` as part of `type_`.

After this, we can write just this.

`decoratorList? accessibilityModifier? identifierOrPattern typeAnnotation?`

File: typescript/TypeScriptParser.g4 [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+    | bindingPattern
+    ;
+
+optionalParameterList: optionalParameter (',' optionalParameter)* ;
+
+optionalParameter
+    : decoratorList? accessibilityModifier? identifierOrPattern '?' typeAnnotation?
+    | decoratorList? accessibilityModifier? identifierOrPattern typeAnnotation? initializer
+    | decoratorList? Identifier '?' ':' StringLiteral
+    ;
```
[the remark this answers is not in the document]
Comment (loonydev) [reply typescript/TypeScriptParser.g4]: After moving `StringLiteral` as part of `type_`, we can ignore part
`Identifier '?' ':' StringLiteral`.

So just
`decoratorList? ( accessibilityModifier? identifierOrPattern ('?' typeAnnotation? | typeAnnotation? initializer))`

File: typescript/TypeScriptParser.g4 (hunk 5 above)
Comment (loonydev) [reply typescript/TypeScriptParser.g4]: I'm not sure that is simplify grammer...

Comment (KvanTTT) [reply typescript/TypeScriptParser.g4]: Ok, it's not necessary.

**Comment by KvanTTT** [comment]
> Also I'm going to add targets from JS grammer later (Go, CShapr, Cpp, JS).

So, should we merge this PR or should await for abovementioned to complete?

**Comment by loonydev** [comment]
Yes, you can merge. I haven't time right now, maybe somebody do it faster

**Comment by studentmain** [comment]
FYI: #1553 Updated JS grammar to current version, you can port them into TS grammar

**Comment by loonydev** [comment]
@studentmain I saw that you update JS to ES2020 version, thx a lot,  but TS based on ES6 version. So, i'm not sure about integration new JS grammar into TS.

**Pull request merged**